src/query_sink.rs

```diff
@@ -127,73 +107,66 @@ impl Stream for AsyncQueryResultStream {
-    impl IWbemObjectSink for QuerySink {
-        unsafe fn indicate(
-            &self,
-            lObjectCount: c_long,
-            apObjArray: *mut *mut IWbemClassObject
-        ) -> HRESULT {
-            trace!("Indicate call with {} objects", lObjectCount);
-            // Case of an incorrect or too restrictive query
-            if lObjectCount <= 0 {
-                return WBEM_S_NO_ERROR as i32;
-            }
+/// Implementation for [IWbemObjectSink](https://docs.microsoft.com/en-us/windows/win32/api/wbemcli/nn-wbemcli-iwbemobjectsink).
+/// This [Sink](https://en.wikipedia.org/wiki/Sink_(computing))
+/// receives asynchronously the result of the query, through Indicate calls.
+/// When finished,the SetStatus method is called.
+/// # <https://docs.microsoft.com/fr-fr/windows/win32/wmisdk/example--getting-wmi-data-from-the-local-computer-asynchronously>
+impl IWbemObjectSink_Impl for QuerySink {
+    fn Indicate(&self, lObjectCount: i32, apObjArray: *const Option<IWbemClassObject>) ->  WinResult<()> {
+        trace!("Indicate call with {} objects", lObjectCount);
+        // Case of an incorrect or too restrictive query
+        if lObjectCount <= 0 {
+            return Ok(());
+        }
 
-            let lObjectCount = lObjectCount as usize;
-            let mut res = WBEM_S_NO_ERROR as i32;
-
-            // The array memory of apObjArray is read-only
-            // and is owned by the caller of the Indicate method.
-            // IWbemClassWrapper::clone calls AddRef on each element
-            // of apObjArray to make sure that they are not released,
-            // according to COM rules.
-            // https://docs.microsoft.com/en-us/windows/win32/api/wbemcli/nf-wbemcli-iwbemobjectsink-indicate
-            // For error codes, see https://docs.microsoft.com/en-us/windows/win32/learnwin32/error-handling-in-com
-            self.stream
-                .extend((0..lObjectCount).map(|i| {
-                if let Some(p_el) = NonNull::new(*apObjArray.add(i)) {
-                    let wbemClassObject = unsafe {
-                        IWbemClassWrapper::clone(p_el)
-                    };
-
-                    Ok(wbemClassObject)
-                } else {
-                    res = E_POINTER;
+        let lObjectCount = lObjectCount as usize;
+        let mut res = Ok(());
+
+        // The array memory of apObjArray is read-only
+        // and is owned by the caller of the Indicate method.
+        // IWbemClassWrapper::clone calls AddRef on each element
+        // of apObjArray to make sure that they are not released,
+        // according to COM rules.
+        // https://docs.microsoft.com/en-us/windows/win32/api/wbemcli/nf-wbemcli-iwbemobjectsink-indicate
+        // For error codes, see https://docs.microsoft.com/en-us/windows/win32/learnwin32/error-handling-in-com
+        let objs = unsafe {
+            std::slice::from_raw_parts(
```

Comment:
The reason the old code used `ptr.add` was that `from_raw_parts` has a bunch of safety reqs which I didn't want to risk violating.
If you think that the `apObjArray` is OK to use in this context, that's fine by me 🙂

Reply:
I have updated the safety comment, imho it's "safe" to do so, as those safety requirements are supposed to be guaranteed by the API for Indicate (and if they are not upheld, we would be UB anyway as any read from `apObjArray + i` would be OOB)

Reply:
I think the note that `The array memory itself is read-only` from MSDN is the one thing I wasn't sure about 👍